### PR TITLE
testutil: Add ScrapeAndCompare

### PR DESCRIPTION
**Pull request by sazary**
As discussed in #993 by @bwplotka, it would be nice to have a utility function that scrapes an exporter and compares it to a set of expected values.  
I've mostly copied the approach used in https://github.com/kubernetes/kubernetes/pull/108277 and refactored some minor functions in the `testutil` package.

Signed-off-by: sazary <soroosh@azary.ir>

### CODE REVIEW

**Comment by sazary** [comment]
hey @bwplotka let me know what you think about this

if this is okay, could you please approve it's workflows?

thanks

File: prometheus/testutil/testutil.go [hunk 1]
```diff
@@ -40,13 +40,12 @@ package testutil
...
 	"bytes"
 	"fmt"
-	"io"
-	"reflect"
-
 	"github.com/davecgh/go-spew/spew"
-	"github.com/prometheus/common/expfmt"
-
 	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
```
Comment (dohnto) [inline prometheus/testutil/testutil.go:45]: Nit: I believe the ordering and grouping imports has a reasoning (first group is stdlib, second group is 3rd parties and third group is "local imports") and shoudn't be changed without a solid reason.

[goimports](https://pkg.go.dev/golang.org/x/tools/cmd/goimports#pkg-overview) is a tool, that might automates this.

File: prometheus/testutil/testutil.go [hunk 2]
```diff
@@ -153,6 +152,33 @@ func GatherAndCount(g prometheus.Gatherer, metricNames ...string) (int, error) {
...
+// ScrapeAndCompare calls a remote exporter's endpoint which is expected to return some metrics in
+// plain text format. Then it compares it with the results that the `expected` would return.
+// If the `metricNames` is not empty it would filter the comparison only to the given metric names.
+func ScrapeAndCompare(url string, expected io.Reader, metricNames ...string) error {
+	resp, err := http.Get(url)
+	if err != nil {
+		return fmt.Errorf("scraping metrics failed: %s", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (dohnto) [inline prometheus/testutil/testutil.go:164]: Shoudn't we ensure that we close the body here?
```suggestion
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
```

File: prometheus/testutil/testutil.go [hunk 3]
```diff
@@ -182,17 +208,35 @@ func TransactionalGatherAndCompare(g prometheus.TransactionalGatherer, expected
...
+
+// convertReaderToMetricFamily would read from a io.Reader object and convert it to a slice of
+// dto.MetricFamily.
+func convertReaderToMetricFamily(reader io.Reader) ([]*dto.MetricFamily, error) {
 	var tp expfmt.TextParser
-	wantRaw, err := tp.TextToMetricFamilies(expected)
+	notNormalized, err := tp.TextToMetricFamilies(reader)
 	if err != nil {
-		return fmt.Errorf("parsing expected metrics failed: %s", err)
+		return nil, fmt.Errorf("converting reader to metric families failed: %s", err)
```
Comment (dohnto) [inline prometheus/testutil/testutil.go:226]:
```suggestion
		return nil, fmt.Errorf("converting reader to metric families failed: %w", err)
```

**Comment by sazary** [comment]
@dohnto thanks for your review

I applied your first & second suggestions and amended them to my previous commit, 6f41614

About the third one, I did it in a separate commit 86f0b14 and applied it to any occurrence in this file

**Review by kakkoyun — APPROVED** [review]
LGTM!

Anything to add @bwplotka ?

**Comment by kakkoyun** [comment]
@sazary Is it possible to rebase and fix conflicts?

**Comment by kakkoyun** [comment]
@sazary I've fixed the conflicts for you ;) It's good to merge now.

**Review by bwplotka — APPROVED** [review]
Thanks a lot!

**Pull request merged**